### PR TITLE
Disable sandboxing for macOS tests in CI

### PR DESCRIPTION
**Pull request by vashworth**
macOS 14 added new requirements that un-codesigned sandbox apps must be granted access when changed. Waiting for this UI caused macOS tests to fail on macOS 14. Additionally, adding codesigning is not sufficient, since it must still be approved before codesigning is enough to pass the check. As a workaround, this PR disables sandboxing for macOS tests in CI.

Screenshot 2024-05-30 at 2 41 33 PM

https://developer.apple.com/documentation/updates/security#June-2023)
> App Sandbox now associates your macOS app with its sandbox container using its code signature. The operating system asks the person using your app to grant permission if it tries to access a sandbox container associated with a different app. For more information, see [Accessing files from the macOS App Sandbox](https://developer.apple.com/documentation/security/app_sandbox/accessing_files_from_the_macos_app_sandbox).

And that link explains why this is happening on a macOS 14 update:

> In macOS 14 and later, the operating system uses your app’s code signature to associate it with its sandbox container. If your app tries to access the sandbox container owned by another app, the system asks the person using your app whether to grant access. If the person denies access and your app is already running, then it can’t read or write the files in the other app’s sandbox container. If the person denies access while your app is launching and trying to enter the other app’s sandbox container, your app fails to launch.
> 
> The operating system also tracks the association between an app’s code signing identity and its sandbox container for helper tools, including launch agents. If a person denies permission for a launch agent to enter its sandbox container and the app fails to start, launchd starts the launch agent again and the operating system re-requests access.

Fixes packages part of https://github.com/flutter/flutter/issues/149264.

Verified tests pass:
https://ci.chromium.org/ui/p/flutter/builders/staging.shadow/Mac_arm64%20macos_platform_tests%20master%20-%20packages/6/overview

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [ ] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or this PR is [exempt from CHANGELOG changes].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version
[following repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[exempt from CHANGELOG changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

### CODE REVIEW

File: packages/pigeon/tool/shared/native_project_runners.dart [hunk 1]
```diff
@@ -38,11 +44,23 @@ Future<int> runFlutterBuild(
...
+  String platform, {
   String? sdk,
   String? destination,
+  String? configuration,
   List<String> extraArguments = const <String>[],
 }) {
+  File? disabledSandboxEntitlementFile;
+  if (extraArguments.contains('test') &&
+      platform.toLowerCase() == 'macos' &&
+      Platform.environment['LUCI_CI'] == 'True') {
```
Comment (vashworth) [inline packages/pigeon/tool/shared/native_project_runners.dart:57]: @tarrinneal Do you think we should check if we're in CI here or just always disable sandboxing when running `xcodebuild test` on macOS? If we do it only in CI, hypothetically if you're running it locally, you'll see the pop up dialog and need to accept it.

Comment (stuartmorgan) [reply packages/pigeon/tool/shared/native_project_runners.dart:57]: I would rather we not special-case CI when we can avoid it; differences between how tests run locally vs CI can make it frustrating for people to debug CI failures.

File: packages/pigeon/tool/shared/native_project_runners.dart [hunk 2]
```diff
@@ -52,12 +70,67 @@ Future<int> runXcodeBuild(
...
+    originalEntitlementFileContents.replaceAll(
+      RegExp(r'<key>com\.apple\.security\.app-sandbox<\/key>[\S\s]*?<true\/>'),
+      '''
+<key>com.apple.security.app-sandbox</key>
+	<false/>''',
+    ),
+  );
+
+  return disabledSandboxEntitlementFile;
+}
```
Comment (stuartmorgan) [inline packages/pigeon/tool/shared/native_project_runners.dart:132]: And given that I think we could remove all this code and just check in the change to the file. This is a test-only app, and what it's designed to test shouldn't be affected by the sandbox.

File: script/tool/lib/src/common/xcode.dart [hunk 3]
```diff
@@ -156,4 +166,52 @@ class Xcode {
...
+
+  /// Finds and copies macOS entitlements file. In the copy, disables sandboxing.
+  /// If entitlements file is not found, returns null.
+  ///
+  /// As of macOS 14, testing a macOS sandbox app may prompt the user to grant
+  /// access to the app. To workaround this in CI, we create and use a entitlements
+  /// file with sandboxing disabled. See
+  /// https://developer.apple.com/documentation/security/app_sandbox/accessing_files_from_the_macos_app_sandbox.
+  File? _createDisabledSandboxEntitlementFile(
+    Directory macosDirectory,
```
Comment (stuartmorgan) [inline script/tool/lib/src/common/xcode.dart:178]: Nit: `macOSDirectory`, per Dart style rules.

File: script/tool/lib/src/common/xcode.dart [hunk 4]
```diff
@@ -156,4 +166,52 @@ class Xcode {
...
+    String configuration,
+  ) {
+    String entitlementDefaultFileName;
+    if (configuration == 'Release') {
+      entitlementDefaultFileName = 'Release';
+    } else {
+      entitlementDefaultFileName = 'DebugProfile';
+    }
```
Comment (stuartmorgan) [inline script/tool/lib/src/common/xcode.dart:186]: Prefer conditionals for things like this to allow for a single final assignment and no chance of divergent logic:

`final String entitlementDefaultFileName = configuration == 'Release' ? 'Release' : 'DebugProfile';`

File: packages/pigeon/tool/shared/native_project_runners.dart (hunk 2 above)
Comment (vashworth) [reply packages/pigeon/tool/shared/native_project_runners.dart]: Sounds good, done

**Review by stuartmorgan — APPROVED** [review]
LGTM

**Pull request merged**